Pointer Compression blog post

santiaboy opened the pull request:
Pointer Compression deep dive blog post

santiaboy commented:
There is a todo in the markdown file, since I am not sure if we can do that (and if so, how to do it).
The rest of the file is ready for publishing

mathiasbynens commented on src/blog/pointer-compression.md:
```diff
@@ -0,0 +1,478 @@
+In turn, it reduces Chrome’s renderer process memory by 20% on Desktop and 10% in Mobile. This difference appears due to V8 being a lower percentage of the renderer process memory on Mobile.
+
+<figure>
+  <img src="/_img/pointer-compression/V8-heap-memory.svg" width="600" height="371" alt="Memory savings when browsing in Windows 10" loading="lazy">
```
`V8-heap-memory.svg` → `v8-heap-memory.svg`

santiaboy replied:
Done!

RReverser commented:
Please fix issues marked by markdownlint - you can see them either in Travis CI logs or by running `npm run lint` locally.

RReverser commented:
@santiaboy Ping.

santiaboy commented:
There are still some svgs to be updated and minified, but I had a question about a lint failure.
We want to show code inside a table but it is complaining saying "no-inline-html Inline HTML [Element: br]". How do we solve this?

RReverser commented:
@santiaboy Hmm yeah for that one you'd probably want to just disable the lint around the table using

```
<!-- markdownlint-disable no-inline-html -->
...
<!-- markdownlint-enable no-inline-html -->
```

RReverser commented:
Update: Turns out we already use a plugin that allows for multiline cells, so I enabled the corresponding option, transformed first table in the post to use it as an example, and fixed style issue for code blocks in a table.

Transforming the rest following this one as an example should be fairly straightforward (even though I'm not saying "easy", because syntax is a bit weird and not supported by standard table generators unfortunately).

Let me know if you run into any issues!

RReverser commented:
Btw, I suggest to use ```` ```asm ```` on other code blocks in the post as well - currently they don't seem to use any highlighting even outside the tables, e.g.

![image](https://user-images.githubusercontent.com/557590/77570618-fd906c00-6ec3-11ea-8ddb-067c6194bc06.png)

RReverser commented on src/blog/pointer-compression.md:
````diff
@@ -0,0 +1,514 @@
+movlsxlq rax, <mem>
+```
+
+swe got yet another 2% improvement.
````
Looks like a typo. Not sure but was this supposed to be

```suggestion
so got yet another 2% improvement.
```

?

RReverser commented on src/blog/pointer-compression.md:
````diff
@@ -0,0 +1,514 @@
+
+While we were looking at the generated code, we noticed that the decompression of a value that had just been loaded produced code that was a bit too verbose:
+
+```
````
```suggestion
```asm
```

RReverser commented on src/blog/pointer-compression.md:
````diff
@@ -0,0 +1,514 @@
+
+Once we fixed that to sign extend the value loaded from memory directly:
+
+```
````
```suggestion
```asm
```

santiaboy commented:
I updated the tabled with the new version.
There is still the matter of the first graph that I am updating, but I wanted to push since I couldn't get rid of the npm lint errors. It is throwing for some reason.

RReverser commented on src/blog/pointer-compression.md:
````diff
@@ -206,24 +207,40 @@ Let’s take a look at the x64 assembly.
-|               | addq r11,r10            | ```                          \
-|               | ```                     |                              |
+|               | addq r11,r10            |  ```                         \
+|               | ```                     |  <br>                        |
````
Hmm what happened here / why does this need `<br>` again?

RReverser replied:
(same for other tables - we shouldn't need it anymore)

santiaboy replied:
br is there so that both code blocks start at the top

RReverser replied:
Hmm I'm conflicted about it, but I guess fine. Just as another option, would leaving a newline inside the code block help? (then both code blocks would have the same size, which might be even better visually for showing the difference)

santiaboy replied:
That's what I tried first, but it does not work (the new line is removed)

RReverser replied:
Hmm it works on my side. To be clear - I'm talking about an extra newline before the closing ```` ``` ````.

RReverser replied:
@santiaboy Ping - what do you think?

RReverser replied:
Nvm, this one doesn't work.

RReverser replied:
Found how this one worked - you just need to use `| \` in the end of the line with spaces instead of `\`.

@santiaboy can you please try again?

RReverser commented on src/blog/pointer-compression.md:
````diff
@@ -322,12 +339,25 @@ Also, since we don’t care about sign extending the Smi anymore, this change al
+| Decompression | Branchful                    | Smi-corrupting               |
+|---------------|------------------------------|------------------------------|
+| Code          | ```asm                       | ```asm                       \
+|               |   movsxlq r11,[...]          |   movl r11,[rax+0x13]        \
````
There seems to be extra unnecessary spacing in the beginning of each code block line here, unlike in others.

santiaboy replied:
Done

santiaboy commented:
Updated the graph. If we can fix the linter issue we are set.
It is happening since we introduced the multiline code in the table. It is throwing with "Error: Property 'range' of onError parameter is incorrect."

RReverser commented on src/blog/pointer-compression.md:
```diff
@@ -0,0 +1,544 @@
+| Summary       | 20 bytes                | 13 bytes                     \
+|               | 6 instructions executed | 3 or 4 instructions executed \
+|               | no branches             | 1 branch                     \
+|               | 1 additional register   | <br>                         |
```
Wait, so you want all the table cells to be aligned to the top, not just cells with code?

If so, we could add a table class (similarly to `table-wrapper`) that would do just that without any `<br>`s, but it makes table style inconsistent across the blog, as all the other ones are aligned by the middle.

Is there any particular reason you want that alignment?

@mathiasbynens do you have thoughts on this as well maybe?

mathiasbynens replied:
No opinion, other than the usual "let's keep it as simple as possible". I trust you'll make the right decision here 👍

santiaboy replied:
Yes, aligning to the top is what we want. Both sides start with loading and then they differ. It is easier to see the difference(e.g in code size) if they are both aligned at the top

RReverser replied:
> No opinion, other than the usual "let's keep it as simple as possible". I trust you'll make the right decision here 👍

I just meant an opinion on concistency. But if you don't have one, that's fine.

RReverser replied:
@santiaboy I see. If the goal is to show clear matching between results, I wonder if it would make more sense to use proper cells (and rowspan for summary here) then?

Here's what it would look like:
![image](https://user-images.githubusercontent.com/557590/77646970-814a6700-6f5d-11ea-90d0-05537a3b42af.png)

You can check table in `v8-release-80.md` for markdown code example.

mathiasbynens replied:
Strong +1 to this suggestion. This seems perfect!

mathiasbynens replied:
Btw, this is the first time we're adding `<pre>` within tables on v8.dev, so let's make sure this works well on narrow viewports etc. before landing.

santiaboy replied:
Yes, that looks perfect.

RReverser replied:
> within tables on v8.dev, so let's make sure this works well on narrow viewports etc. before landing.

Yeah I double-checked that already, table-wrapper seems to work fine here too and just introduces horizontal scroll when necessary.

RReverser commented:
Added a custom hacky way (or, in fancy words, a CSS4 feature polyfill!) to render table cells with just code inside a bit differently.

Before:
![image](https://user-images.githubusercontent.com/557590/77679702-d18bee00-6f8a-11ea-91c7-751ce5c63df3.png)


After:
![image](https://user-images.githubusercontent.com/557590/77679647-c20ca500-6f8a-11ea-8e9a-1c0956fc56e0.png)

mathiasbynens commented:
@rreverser Looking slick!

RReverser commented:
@mathiasbynens I was wondering whether you'll say that or "oh nooo this is way too hacky" 😅

mathiasbynens commented:
> @mathiasbynens I was wondering whether you'll say that or "oh nooo this is way too hacky" 😅

Not at all :) I’m totally down with this. As long as we can hide the “hackiness” in the build process as opposed to paying the cost on the client-side, I consider it a win!

RReverser commented on src/blog/pointer-compression.md:
```diff
@@ -0,0 +1,537 @@
+Before diving into the implementation, we need to know where we are standing to correctly assess the situation. To measure our memory and performance we use a set of [web pages](https://v8.dev/blog/optimizing-v8-memory) that reflect popular real-world websites. The data showed that V8 contributes up to 60% of Chrome’s [renderer process](https://www.chromium.org/developers/design-documents/multi-p ...
+
+<figure>
+  <img src="/_img/pointer-compression/memory-chrome.svg" width="600" height="371" alt="V8 memory consumption percentage in Chrome’s renderer memory" loading="lazy">
```
Here and on other images: set `alt` to empty with `alt=""`. When we use `figcaption`, there is no point in duplicating same text inside the `alt`, because it will only worsen UX for users on low-speed connections or impaired users, who will now see or hear same text twice.

santiaboy replied:
Done.

RReverser commented on src/blog/pointer-compression.md:
```diff
@@ -0,0 +1,537 @@
+While implementing the Decompression Elimination in TurboFan we learned a lot. The explicit Decompression/Compression node approach had the following properties:
+
+Pros:
+    - Explicitness of such operations allowed us to optimize unnecessary decompressions by doing canonical pattern matching of sub-graphs.
```
These lists don't render as actual lists, you might want to add a newline before the first item.

RReverser commented on src/blog/pointer-compression.md:
```diff
@@ -0,0 +1,542 @@
+## Compressed tagged values and new heap layout
+
+With Pointer Compression, our goal is to somehow fit both kinds of tagged values into 32 bits on 64-bit architectures. We can fit pointers into 32 bits by:
+    - making sure all V8 objects are allocated within a 4GB memory range
```
@santiaboy Looks like this one got missed and still doesn't render like list (btw, the indentation is also a bit excessive).

santiaboy replied:
Done.
The indentation was due to npm run lint. I was originally using " *" and I guess that confused it because it sugested using "-" and 4 spaces.
